enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,232 @@
+
+#### Story 3:
+
+We desire to extend our team responsibilities on upstream cloud providers in the future and gain weight in promoting features into upstream. Getting such for kubernetes repository is harder for us and maintainers, as this would mean giving weight for approving and merging features for the parts of the kubernetes project, not necessarily laying under our responsibilities and scope of knowledge.
+
+#### Story 4:
+
+We’d like to discuss technical details related to a specific cloud in a SIG meeting with people who are also involved into development in this domain, and that way gain useful insights into the cloud infrastructure, improve the overall quality of our features, stay on top of the new features, and improve the relations with maintainers outside of our company, which nevertheless share with us a common goal.
+
+
+### Implementation Details
+
+The `external` `cloud-controller-manager` implementation will be hosted within openshift repository in https://github.com/openshift/cluster-cloud-controller-manager-operator. That repository will manage `CCM` provisioning for metioned cloud providers + `Openstack` out-of-tree provider implementation described in the [document](https://github.com/openshift/enhancements/tree/master/enhancements/cloud-controller-manager/openstack-cloud-controller-manager.md)
+
+`Cloud-controller` will be provisioned in the `openshift-cloud-controller-manager` namespace.
+
+Preparation steps, common for all providers:
+
+- Add support for recognition of the `external` option in [library-go](https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go) which will disable config parsing and add support for `external` field in the `Infrastructure` resource.
+- Port the templates of `cloud-controller` from upstream repositories into proposed https://github.com/openshift/cluster-cloud-controller-manager-operator repository. Those templates will belong to the newly created `openshift-cloud-controller-manager` namespace and be managed with an operator.
+- Operator will be deployed in the `openshift-cloud-controller-manager-operator` namespace.
+
+Each provider implementation will need to do the following minimal set of actions in order to make out-of-tree implemetation work:
+
+- Use `external` flag in the `kube-controller-manager` pod instead of the previously used provider. Remove the `cloud-config` option from the template.
+- Use `external` flag in the `kubelet` pod. Remove the `cloud-config` option from the template.
+- Provision a `cloud-controller-manager` `DaemonSet` in the `openshift-cloud-controller-manager` namespace.
```

Comment:
Why a daemonSet? I suggest a deployment unless there's a reason for us to use static pod, i.e if we need it for the cluster to bootstrap or come back up.

Reply:
Not even a DaemonSet is about static pods.

My questions:
- can we disaster-recover a cluster without ccm running?
- what is non-functional in a cluster during bootstrapping until ccm is up?
- who talks to ccm? Only kcm or also ks and ka?
- what are the version skew compatibility constraints? Can n+1 kcm talk to n ccm and conversely?
- does ccm provide an API? How is it hosted? https endpoint? Through load balancer?
- what are the thoughts about certificate management?

Reply:
> can we disaster-recover a cluster without ccm running?

Without CCM, any new kubelet remains tainted as uninitialised, from the docs:
>> kubelets specifying --cloud-provider=external will add a taint node.cloudprovider.kubernetes.io/uninitialized with an effect NoSchedule during initialization. This marks the node as needing a second initialization from an external controller before it can be scheduled work. Note that in the event that cloud controller manager is not available, new nodes in the cluster will be left unschedulable. The taint is important since the scheduler may require cloud specific information about nodes such as their region or type (high cpu, gpu, high memory, spot instance, etc).

So this would affect recovery if new nodes are needed to be booted

> what is non-functional in a cluster during bootstrapping until ccm is up?

Based on the above, pods will not be able to schedule unless they tolerate the uninitialised taint, while we could do this, without initialised nodes, we don't have any availability information and as such affinities based scheduling decisions are not possible

The above makes me feel like it's a pretty fundamental controller and lean towards running the CCM as a static pod in a similar way to KCM, I'm not entirely sure of the implications of doing this though so would be keen to know if there have previously been discussions about what warrants running as a static pod

Reply:
@Danil-Grigorev can you answer the other questions in the list? They are fundamental to asses possible architectures.

Reply:
More questions:
- what happens if the kcm leader has no access to ccm? Can it continue its work? Will it give up leadership?
- how does a kubelet talk to the ccm? Does every node need a ccm? What about worker nodes?

Reply:
Hey! I tried to answer all the questions in the FAQ section of the proposal. Please unresolve it if you think it is necessary